### ISSUE TITLE
[logind]: suspend-then-hibernate: if hibernation is not available, but s-t-h is set for HandleLidSwitch=, still suspend

### ISSUE THREAD

**Issue by shalzz** [systemd/systemd#10558]
**systemd version the issue has been seen with**
> `239.2-1`

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > System suspends even if it is not able to hibernate

**Unexpected behaviour you saw**
 > 
Suspend fails with suspend-then-hibernate preemptively if hibenation is not possible.

Also I'm not sure hibernation would fail in this case with image compression.

Trying to just suspend works:
```
$ busctl call org.freedesktop.login1 /org/freedesktop/login1 org.freedesktop.login1.Manager CanSuspend
s "yes"

shalzz at shalzz-galp2 in ~/dev/serverless/academics-api on master [!$]
$ busctl call org.freedesktop.login1 /org/freedesktop/login1 org.freedesktop.login1.Manager CanHibernate
s "yes"

shalzz at shalzz-galp2 in ~/dev/serverless/academics-api on master [!$]
$ systemctl suspend
```

But with the `suspend-then-hibernate`
```
$ systemctl suspend-then-hibernate
Failed to put system into suspend followed by hibernate via logind: Not enough swap space for hibernation
```
Swap space: 
```
$ swapon
NAME      TYPE      SIZE USED PRIO
/dev/dm-2 partition   8G 1.6G   -2
```

Ram:
```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:           19Gi       7.3Gi       2.2Gi       1.4Gi        10Gi        10Gi
Swap:         8.0Gi       1.6Gi       6.4Gi
```

logind.conf:
```
#NAutoVTs=6
#ReserveVT=6
#KillUserProcesses=no
#KillOnlyUsers=
#KillExcludeUsers=root
#InhibitDelayMaxSec=5
HandlePowerKey=suspend-then-hibernate
HandleSuspendKey=suspend-then-hibernate
#HandleHibernateKey=hibernate
HandleLidSwitch=suspend-then-hibernate
HandleLidSwitchExternalPower=lock
#HandleLidSwitchDocked=ignore
#PowerKeyIgnoreInhibited=no
#SuspendKeyIgnoreInhibited=no
#HibernateKeyIgnoreInhibited=no
#LidSwitchIgnoreInhibited=yes
#HoldoffTimeoutSec=30s
#IdleAction=ignore
#IdleActionSec=30min
#RuntimeDirectorySize=10%
#RemoveIPC=yes
#InhibitorsMax=8192
#SessionsMax=8192
#UserTasksMax=33%
```

**Steps to reproduce the problem**
Have a lot of apps open or low swap space.

**Comment by poettering** [comment]
not following? what's the bug supposed to be here? are you suggesting we shall implicitly fall back to regular suspend if suspend-then-hibernate is requested but hibernation not available? why should we do that? it sounds more appropriate the request suspend in that case directly?

Not sure i follow?

**Comment by shalzz** [comment]
Hi!
@poettering In my case hibernation is available and works generally but systemd may fail to hibernate if it decides that there's not enough swap space.
The bug here is two fold:
* The systemd algorithm for hibernation simply checks for required space and available space (`r = act <= (size - used)`) , not accounting for image compression. From the [Arch wiki][0] and [kernel documentation][1]:
> Even if your swap partition is smaller than RAM, you still have a big chance of hibernating successfully. According to kernel documentation:
   > ` /sys/power/image_size` controls the size of the image created by the suspend-to-disk mechanism. It can be written a string representing a non-negative integer that will be used as an upper limit of the image size, in bytes. The suspend-to-disk mechanism will do its best to ensure the image size will not exceed that number. However, if this turns out to be impossible, it will try to suspend anyway using the smallest image possible. In particular, if "0" is written to this file, the suspend image will be as small as possible. Reading from this file will display the current image size limit, which is set to 2/5 of available RAM by default.
* If the system cannot genuinely hibernate for whatever reason but the system supports it and has successfully hibernated in the past then `suspend-to-hibernate` should not fail to at the very least suspend-to-ram than not suspending at all since that leads to a lot more loss of power. Especially in situations where the lid close behaviour is set to `HandleLidSwitch=suspend-then-hibernate` and the user is expecting the system to not stay awake.

I hope that makes it clearer.

[0]: https://wiki.archlinux.org/index.php/Power_management/Suspend_and_hibernate#About_swap_partition.2Ffile_size
[1]: https://www.kernel.org/doc/Documentation/power/interface.txt

**Comment by shalzz** [comment]
That said, 
I'm not sure what the right solution would be for the first point but it would be great if systemd does not pre-emptively fails hibernation. Maybe try creating a compressed hibernation image before hand and then compare relevant sizes? Or maybe just for `suspend-then-hibernate` let the kernel handle hibernation and any hibernation failures and does perform any check before hand.

For the second, yes with suspend-then-hibernate, suspending and hibernating should be evaluated and performed independently, implied in the name that the system suspends _then_ hibernates with suspend not failing because of hibernation.

**Comment by poettering** [comment]
Generally, figuring out whether hibernation is likely possibly or not is black magic, it's a heuristic. And I think we should be defensive there and rather shoot too low than too high (i.e. rather say we don't support it if there's any doubt). Note that there's a workaround against these checks, if you know what you do and are sure you know better than our estimations: set the env var  $SYSTEMD_BYPASS_HIBERNATION_MEMORY_CHECK=1 for logind. if you do, then logind won't do it's estimations anymore, and just assume it'll fit.

Now, the second issue, you are definitely right: we should always fallback to plain suspend if suspend-then-hibernate is chosen but doesn't work. We already did some work in that direction (see #10504), but I figure there might be more work to do.

**Comment by poettering** [comment]
Fix pending in #10814, we'll now automatically fall back to plain suspend if necessary.
